explorer/ast/element.h

```diff
@@ -29,21 +29,34 @@ struct NamedValue {
     return f(name, value);
   }
 
+  inline friend auto operator==(const NamedValue& lhs, const NamedValue& rhs)
+      -> bool {
+    return lhs.name == rhs.name && lhs.value == rhs.value;
+  }
+  inline friend auto hash_value(const NamedValue& value) -> llvm::hash_code {
+    return llvm::hash_combine(value.name, value.value);
+  }
```

Comment:
You're providing this pair of functions for various types that are `Decompose`able. Perhaps instead we should be providing blanket support for `==` and hashing for all `Decompose`able types, and adding `Decompose` support to the other cases in this patch where you are currently adding an explicit `==` / `hash_value`?

Reply:
Possibly. I see a few concerns with that approach:
- Where do we put that blanket support, so that it's available whenever any `Decompose`able type is?
- We'd need to use a wrapper type to work around the absence of a `hash_value` overload for `std::vector`, and it would need to be a slightly different wrapper than the one `Arena` uses (because here we want to wrap a reference rather than a copy).
- Those overloads seem like they will be an attractive nuisance in a lot of cases. If you're applying `==` to two `Value`s, you have either a performance bug (you should be comparing their addresses instead) or a correctness bug (you should be calling `ValueEqual` instead), so it seems good that such code currently doesn't compile. More broadly, `==` and hashing are typically not a good fit for types where identity is meaningful, so it would be kind of awkward to add those operations to `Value` as part of a change that makes `Value` identity more meaningful.

Reply:
> * Where do we put that blanket support, so that it's available whenever any `Decompose`able type is?

Perhaps we can put it into `LlvmHasher` (and add an equality comparison function object for the `==` side of things). I think that would also address your other concerns: we can put the vector hashing support in `LlvmHasher`, and not add these `operator==` / `hash_value` functions that may be an attractive nuisance.

Reply:
OK, I've done that, but it's not pretty. The code could probably be cleaned up to some extent, but a fair amount of the extra complexity seems inherent. In addition to the complexity of the metaprogramming itself, we have to reimplement hashing and equality for all the `std` templates that we want to support, in order to ensure that they recurse into our helpers rather than into `hash_value` and `operator==`, because the latter don't support `Decompose`able types.

I would prefer to revert 4362ff1bf6 (the commit implementing this suggestion) but if you strongly prefer it, I could clean it up for merging.

Alternatively, we could probably provide a CRTP helper that implements `operator==` and `hash_value` in terms of `Decompose`. That would require each type to opt in, but it would eliminate most of the per-type boilerplate while avoiding all the extra complexity in `Arena`.

Reply:
I've switched to the CRTP approach, as discussed on Discord.